File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        //sliding window with vector to keep track of last location of particular char.
        vector<int> mp(256,-1);
        int left = 0;
        int right = 0;
        int maxLen = 0;
        while(right< s.size())
        {
            //already seen this character before
            if(mp[s[right]] != -1)
            {
                left = max(left, mp[s[right]] + 1 );
            }
            //update char location.
            mp[s[right]] = right;
            maxLen = max(right-left+1,maxLen);
            right++;
        }
        return maxLen;

    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        int [] chars = new int[128];
        int left = 0;
        int right = 0;
        int r,l;
        while(right < s.length()){
            r = s.charAt(right);
            chars[r] ++;
            
            while(chars[r] > 1 ){
                l = s.charAt(left);
                chars[l] --;
                left++;
            }
            
            ans = Math.max(ans, right - left + 1);
            right ++;
        }
        return ans;
        
    }
}